packages/supplier/src/supplier.ts

```diff
@@ -0,0 +1,181 @@
+import type { Strategy as FreeStrategy, BaseStrategy } from "./strategy";
+import type { Results as FreeResults } from "./results";
+
+/**
+ * Given the definition of specified supplier, create a function that
+ * determines+constructs a supplier strategy for a given set of options.
+ */
+export const forDefinition = <S extends Supplier.Specification>({
+  determineStrategy,
+  strategyConstructors
+}: Supplier.Definition<S>): CreateSupplier<S> => <
+  N extends Supplier.StrategyName<S>
+>(
+  options: Supplier.Strategy.Constructor.Options<S, N>
+): Supplier<S, N> => {
+  const strategyName = determineStrategy(options);
+  const Strategy = strategyConstructors[strategyName];
+
+  // @ts-ignore since we can't figure out N from inside
+  const supplier: Supplier<S, N> = new Strategy(options);
+  return supplier;
+};
+
+/**
+ * A function that constructs a Supplier
+ */
+export type CreateSupplier<S extends Supplier.Specification> = <
+  N extends Supplier.StrategyName<S>
+>(
+  options: Supplier.Strategy.Constructor.Options<S, N>
+) => Supplier<S, N>;
+
+/**
+ * An object of type Supplier<S, N> provides an interface for loading and
+ * possibly listing versions of the supplied component
+ *
+ * @dev for known strategy names, this computes the method signature of
+ *      `load()` according to the corresponding strategy specification;
+ *
+ *      for the default of "any strategy name", this type will resolve to an
+ *      interface that may require the use of defined type guards
+ *      (e.g. `if(supplier.allowsListingVersions()) { supplier.list(); }`)
+ */
+export type Supplier<
+  S extends Supplier.Specification,
+  N extends Supplier.StrategyName<S> = Supplier.StrategyName<S>
+> = Supplier.StrategyName<S> extends N
+  ? BaseStrategy<Supplier.Strategy.Specification<S, N>>
+  : FreeStrategy<Supplier.Strategy.Specification<S, N>>;
+
+export namespace Supplier {
+  /**
+   * Type-level description of a supplier.
+   */
+  export type Specification = {
+    /**
+     * When creating a supplier, the `options` type represents whatever data
+     * is necessary for creating an instance of the specified supplier.
+     *
+     * N.B. that this type is distinct from any strategy-specific options type:
+     * `createSupplier()` requires an options argument that conforms to the
+     * intersection of this `options` type, as well as the options type for at
+     * least one strategy.
+     */
+
+    options: any;
+
+    /**
+     * A supplier specification includes a unified `results` specification,
+     * which defines the `supplier.load()` and `supplier.list()` return types
+     * for all strategies.
+     */
+    results: FreeResults.Specification;
+
+    /**
+     * A supplier specification includes specifications for all strategies by
```

Comment:
I'm not sure what you're suggesting? I'm probably missing a more charitable interpretation of your statement, but I don't think it's correct to say that it's the _set of keys_ that define this; rather, _a given specification_ is the definition, itself... the keys are a part of that definition.

Reply:
Yes, that's a different issue from what I'm talking about.

Your comment currently states, "a supplier specification includes specifications for all strategies by name".  OK.  But what defines what constitutes "all strategies"?  Someone attempting to use this type might misunderstand this, thinking that what constitutes "all strategies" is defined elsewhere -- perhaps as a fixed list defind somewhere -- and go on a wild goose chase trying to find where so they can find out what they are so they can be sure to include them all.

In fact, there is no such external list; it is the set of keys to this field that *defines* what constitutes "all strategies" in this context.  The comment should be clear that this field is *defining* what constitutes "all strategies", rather than the alternate interpretation that it *must include* all strategies, the set of which is defined elsewhere.